app/src/server/tasks/evaluateTestSetEntries.task.ts

```diff
@@ -0,0 +1,428 @@
+      try {
+        let entry;
+        if (result.datasetEvalOutputSource.modelId !== ORIGINAL_MODEL_ID) {
+          entry = await prisma.fineTuneTestingEntry.findUniqueOrThrow({
+            where: {
+              modelId_datasetEntryId: {
+                modelId: result.datasetEvalOutputSource.modelId,
+                datasetEntryId: result.datasetEvalDatasetEntry.datasetEntryId,
+              },
+            },
+            include: {
+              fineTune: true,
+            },
+          });
+          if (isComparisonModel(entry.modelId)) {
+            entryIds.push(getComparisonModelName(entry.modelId));
+          } else if (entry.fineTune) {
+            entryIds.push("openpipe:" + entry.fineTune.slug);
+          } else {
+            throw new Error("No fineTune or comparison model found for entry");
+          }
+        } else {
+          entry = result.datasetEvalDatasetEntry.datasetEntry;
+          entryIds.push("the original model");
+        }
+        entries.push(entry);
+      } catch (e) {
+        console.error("error getting entry for result", result, e);
+        await prisma.datasetEvalResult.updateMany({
+          where: {
+            id: {
+              in: [firstResult.id, secondResult.id],
+            },
+          },
+          data: {
+            status: "ERROR",
+            errorMessage: "Error retrieving relevant input for evaluation",
+          },
+        });
+        return;
+      }
+    }
+
+    const [firstEntry, secondEntry] = entries;
+    const [firstEntryId, secondEntryId] = entryIds;
+
+    if (!firstEntry?.output || !secondEntry?.output || !firstEntryId || !secondEntryId) {
+      await prisma.datasetEvalResult.updateMany({
+        where: {
+          id: {
+            in: [firstResult.id, secondResult.id],
+          },
+        },
+        data: {
+          status: "ERROR",
+          errorMessage: "Error preparing for evaluation",
+        },
+      });
+      return;
+    }
+
+    let explanation;
+    let judgement;
+
+    const instructions = firstResult.datasetEvalDatasetEntry.datasetEval.instructions;
+
+    try {
+      const input: ChatCompletionCreateParams = {
+        model: "gpt-4",
+        messages: [
+          {
+            role: "system",
+            content: `You are an intelligent and fair judge of chatbots. Evaluate the following two chatbot responses and choose which one is better in the following way: ${
+              instructions ?? ""
+            }`,
+          },
+          {
+            role: "user",
+            content: formatDatasetEntryInputInstructions(
+              firstResult.datasetEvalDatasetEntry.datasetEntry,
+            ),
+          },
+          {
+            role: "user",
+            content: `This is what Alice said:\n\n${JSON.stringify(firstEntry.output)}`,
+          },
+          {
+            role: "user",
+            content: `This is what Bob said:\n\n${JSON.stringify(secondEntry.output)}`,
+          },
+          {
```

Comment:
I'd remove this final message unless you've tried and it makes a noticeable positive impact on the quality.

Reply:
Let's experiment without it. I included it because I've heard LLMs pay the most attention to the very beginning and ending of inputs, and I want to make sure that these specific instructions are observed. However, it's possible including them in the system message will be sufficient.

Reply:
It does seem that adding this at the end is important to keeping the judgement focused on the particular criteria the user specifies.